Refactor oneapi packages to align with standalone installers

rscohn2 opened the pull request:
This is a continuation of #21904 . I could not reopen because I did a force push.

* Align spack packages to oneapi standalone packages: https://software.intel.com/content/www/us/en/develop/articles/oneapi-standalone-components.html
* Use the vars.sh to setup the environment
* eliminate a lot of complicated code that centralized generating the URL's. I found it is easier to implement and understand if you cut and paste the versions and URLs from https://software.intel.com/content/www/us/en/develop/articles/oneapi-standalone-components.html into the package.
* add some documentation

I attempted to add macos support for #22178, but ran into some problems--the installer requires sudo and I don't know how to manage that. I removed all the mac code, but saved it in this branch: https://github.com/rscohn2/spack/tree/dev/mac.

@scheibelp: ready for final review. It might be easier to look at the full files then the diffs

rscohn2 committed:
Refactor oneapi packages to align with standalone installers

rscohn2 commented:
@DanVev: Please review

glennpj commented:
@rscohn2 I saw a message on the XSede Campus champions list that there is some question about the legality of distributing OneAPI binaries. Distribution was not really an issue with Parallel Studio because that required a license checkout. I am not sure what the implications of the distribution policies, once clarified, are for Spack and people using Spack to provide packages for end users. I do not have any further details at the moment but wanted to bring that up.

rscohn2 commented:
@scheibelp: Can you review for merge? Someone reported
```
spack load intel-oneapi-compilers
```
is broken, and this PR fixes it.

scheibelp commented on lib/spack/docs/build_systems/inteloneapipackage.rst:
```diff
@@ -0,0 +1,237 @@
+
+
+
+Tips for configuring client packages to use MKL
```
This looks equivalent similar to the corresponding section in `intelpackage.rst`, could you link that section here? More specifically this appears to copy the sub-item `To use MKL as provider for BLAS, LAPACK, or ScaLAPACK`. You could create a link in that file, then reference that link here.

rscohn2 replied:
Removed all redundant content and added a section at the end with pointers into intelpackage.rst.

scheibelp commented on lib/spack/docs/build_systems/inteloneapipackage.rst:
```diff
@@ -0,0 +1,237 @@
+
+
+.. _`Intel installers`: https://software.intel.com/content/www/us/en/develop/documentation/installation-guide-for-intel-oneapi-toolkits-linux/top.html
+.. _`Spack package list`: https://spack.readthedocs.io/en/latest/package_list.html#intel-oneapi-ccl
```
Ideally the rst docs will reference other rst files: otherwise reference errors may not be detected by the sphinx build. `:ref:<package-list>` should work. You could also suggest that users try `spack list -d oneAPI` (which would allow users to then enter things like `spack info x` for each such package.

rscohn2 replied:
done

scheibelp commented on lib/spack/spack/build_systems/oneapi.py:
```diff
@@ -22,46 +24,44 @@ class IntelOneApiPackage(Package):
+        if installer_path is None:
+            installer_path = basename(self.url_for_version(spec.version))
+
+        if platform == 'linux':
```
(no change needed) I was thinking it would be worthwhile to add an "else" that raises an error. That being said, it looks like you only define versions on the `linux` platform, so I think this will cause Spack concretization to fail on other platforms and the `else` would not be reached.

Side note: A `conflicts` statement would be great but Spack doesn't allow you to say something like `conflicts(!linux)` so this is the earliest we could catch the error in a reasonable manner without doing something like adding a conflicts statement for each platform where we anticipate running Spack.

rscohn2 replied:
no change

scheibelp commented on var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:
```diff
@@ -5,40 +5,39 @@
-    def _join_prefix(self, path):
-        return join_path(self.prefix, 'compiler', 'latest', 'linux', path)
+    def _join_prefix(self, p):
+        return path.join(self.prefix, 'compiler/latest/linux', p)
```
(minor) The original call

```
join_path(self.prefix, 'compiler', 'latest', 'linux', path)
```

ensures that the proper file separators are used. As I understand it this is only ever likely to run on systems that use `/` as a file separator, so this isn't a big deal (but it is idiomatic to avoid using separators in calls to `os.path.join`).

rscohn2 replied:
eliminated all / in the file

scheibelp commented on lib/spack/spack/build_systems/oneapi.py:
```diff
@@ -22,46 +24,44 @@ class IntelOneApiPackage(Package):
-                       components,
-                       releases,
-                       url_name):
+    def component_info(self, dir_name):
```
Currently it looks like the only purpose of `dir_name` is to confirm that the component was installed successfully. If that's true, could you update this to be a property

e.g.

```
@property
def component_dir(self):
  raise NotImplementedError
```

the inheritors of the `IntelOneapiPackage` can define this property, and the install method can check

```
if not isdir(join(prefix, self.component_dir)):
```

that approach matches some common patterns in Spack packages when a build system requires a specific detail from a subclass.  Another option is to add a docstring here clarifying the use of `component_info` like

```
def component_info():
  """Define the expected installation directory. After the install script runs,
     IntelOneApiPackage will look for this directory to make sure the install
     was successful.
  """
```

rscohn2 replied:
I like the first solution, but my understanding of python is that a base class cannot reference methods in the derived class. I will add the docstring.

scheibelp replied:
(no change needed) Regarding:

> I like the first solution, but my understanding of python is that a base class cannot reference methods in the derived class.

You can do this if desired in Python (the way I was suggesting doing it was somewhat analogous to defining an abstract class in Java, if that reference point is worth anything). Like I said though it's fine to just add a comment: that clarifies what the function is for.

rscohn2 replied:
You are right. Using the component_dir solution. It improves clarity for the packages.

scheibelp reviewed (changes requested):
This looks good except for a couple minor issues: the most important ones IMO relate to the documentation and I also have a minor refactoring request related to `component_info`. Let me know if you disagree or have questions - thanks!

vdanilin commented on lib/spack/docs/build_systems/inteloneapipackage.rst:
```diff
@@ -0,0 +1,237 @@
+Add the oneAPI compilers to the set of compilers that Spack can use::
+
+  spack compiler add `spack location -i intel-oneapi-compilers`/latest/linux/bin/intel64
+  spack compiler add `spack location -i intel-oneapi-compilers`/latest/linux/bin
```
As `spack location -i intel-oneapi-compilers` return a path to oneAPI root directory, I guess it lacks "compiler" folder:
_> spack location -i intel-oneapi-compilers
/work/spack/opt/spack/linux-centos8-skylake/gcc-8.3.1/intel-oneapi-compilers-2021.1.2-lahorl3lrtyclrowf2psfgsga3ncyvav
\>pwd
  spack compiler add \`spack location -i intel-oneapi-compilers\`/latest/linux/bin/intel64
/work/spack/opt/spack/linux-centos8-skylake/gcc-8.3.1/intel-oneapi-compilers-2021.1.2-lahorl3lrtyclrowf2psfgsga3ncyvav/**compiler**/latest/linux/bin_

```suggestion
  spack compiler add `spack location -i intel-oneapi-compilers`/compiler/latest/linux/bin/intel64
  spack compiler add `spack location -i intel-oneapi-compilers`/compiler/latest/linux/bin
```

rscohn2 committed:
address review comments

rscohn2 committed:
fix compiler add command

Co-authored-by: Vasily Danilin <vasily.danilin@yandex.ru>

rscohn2 commented:
@danvev Thanks for catching the error in the path.

On Wed, Mar 17, 2021 at 8:55 AM Vasily Danilin ***@***.***>
wrote:

> *@DanVev* requested changes on this pull request.
> ------------------------------
>
> In lib/spack/docs/build_systems/inteloneapipackage.rst
> <https://github.com/spack/spack/pull/22202#discussion_r595940517>:
>
> > +  spack compiler add `spack location -i intel-oneapi-compilers`/latest/linux/bin/intel64
>
> +  spack compiler add `spack location -i intel-oneapi-compilers`/latest/linux/bin
>
>
> As spack location -i intel-oneapi-compilers return a path to oneAPI root
> directory, I guess it lacks "compiler" folder:
>
>
>
>
> *> spack location -i intel-oneapi-compilers
> /work/spack/opt/spack/linux-centos8-skylake/gcc-8.3.1/intel-oneapi-compilers-2021.1.2-lahorl3lrtyclrowf2psfgsga3ncyvav
> >pwd spack compiler add spack location -i
> intel-oneapi-compilers/latest/linux/bin/intel64
> /work/spack/opt/spack/linux-centos8-skylake/gcc-8.3.1/intel-oneapi-compilers-2021.1.2-lahorl3lrtyclrowf2psfgsga3ncyvav/compiler/latest/linux/bin*
> ⬇️ Suggested change
>
> -  spack compiler add `spack location -i intel-oneapi-compilers`/latest/linux/bin/intel64
>
> -  spack compiler add `spack location -i intel-oneapi-compilers`/latest/linux/bin
>
> +  spack compiler add `spack location -i intel-oneapi-compilers`/compiler/latest/linux/bin/intel64
>
> +  spack compiler add `spack location -i intel-oneapi-compilers`/compiler/latest/linux/bin
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/22202#pullrequestreview-614207120>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOY2WWCPFTVCOAHNI2U2TLTECRE7ANCNFSM4Y4O3AJQ>
> .
>

scheibelp commented on lib/spack/spack/build_systems/oneapi.py:
```diff
@@ -22,46 +24,52 @@ class IntelOneApiPackage(Package):
+    # oneAPI license does not allow mirroring outside of the
+    # organization (e.g. University/Company).
+    redistribute_source = False
+    
```
According to style checks, this line contains whitespace.

rscohn2 replied:
Fixed all ```spack style``` issues.

scheibelp commented on var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:
```diff
@@ -5,82 +5,77 @@
+            installer_path=glob.glob(path.join('fortran-installer', '*'))[0])
+
+        # Some installers have a bug and do not return an error code when failing
+        if not path.isfile(path.join(prefix, 'compiler', 'latest', 'linux', 'bin', 'intel64', 'ifort')):
```
According to the style checks (https://github.com/spack/spack/pull/22202/checks?check_run_id=2132375022) this line is too long and you'll need to break it up into multiple lines.

rscohn2 replied:
Fixed all ```spack style``` issues.

scheibelp reviewed (changes requested):
This looks good except there are a couple style issues that are triggering the CI failures. I added comments for those.

rscohn2 committed:
resolve style errors, use property for component_dir

rscohn2 commented:
Package tests passed: https://github.com/rscohn2/oneapi-spack-tests/actions/runs/665511333
Still waiting for spack tests.

glennpj commented:
I just noticed that the intel-oneapi-dal package does not have `provides('daal')` set.

glennpj commented:
Should a `dnn` provider be added? I guess there are two packages that provide that, but perhaps they are not compatible.
- onednn
- intel-oneapi-dnn

rscohn2 commented:
> Should a `dnn` provider be added? I guess there are two packages that provide that, but perhaps they are not compatible.
> 
> * onednn
> * intel-oneapi-dnn

@vpirogov: Are onednn product build and open source build compatible? Can an application that builds against one also build against the other?

rscohn2 committed:
add provides for all oneapi packages that have open source equivalents

rscohn2 commented:
> I just noticed that the intel-oneapi-dal package does not have `provides('daal')` set.

I added ```provides(onexxx)``` for all the packages that have open source equivalents and added ```provides(daal)```

rscohn2 committed:
remove newly added provides one*

rscohn2 commented:
@scheibelp all required tests passed. I will do the provides in a separate PR because we need to discuss naming scheme and what will be the default provider. I think we also want to change the default provider for mkl so it will be the oneapi version.

scheibelp commented:
Thanks!
